File: app/src/main/java/com/dumuzeyn/mp3player/LibraryMutationController.java
package com.dumuzeyn.mp3player;

import java.util.HashSet;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.RejectedExecutionException;

/** Coordinates committed SQLite removals with the activity and Media3 projections. */
final class LibraryMutationController implements AutoCloseable {
    private final MainActivityCore host;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private volatile boolean closed;

    LibraryMutationController(MainActivityCore host) {
        this.host = host;
    }

    void removeTrack(Track track) {
        if (track == null) {
            return;
        }
        execute(() -> {
            LibraryMutationStore store = new LibraryMutationStore(host);
            try {
                publish(store.removeTrack(track));
            } finally {
                store.close();
            }
        });
    }

    void removeSource(LibrarySource source) {
        if (source == null) {
            return;
        }
        execute(() -> {
            publish(PersistedFolderStore.forget(host, source));
        });
    }

    void clearLibrary() {
        execute(() -> {
            publish(PersistedFolderStore.clear(host));
        });
    }

    @Override
    public void close() {
        closed = true;
        executor.shutdownNow();
    }

    private void execute(Runnable mutation) {
        try {
            executor.execute(() -> {
                try {
                    mutation.run();
                } catch (RuntimeException error) {
                    VoltuneLog.failure("library_mutation_failed", error);
                }
            });
        } catch (RejectedExecutionException ignored) {
            // Activity is already closing.
        }
    }

    private void publish(RemovedLibraryItems removed) {
        LibraryMutationClock.advance();
        for (LibrarySource source : removed.sources) {
            PersistedFolderStore.releaseReadPermission(host, source.asUri());
        }
        host.uiHandler.post(() -> applyToUi(removed));
    }

    private void applyToUi(RemovedLibraryItems removed) {
        if (closed) {
            return;
        }
        if (removed.clearQueue) {
            host.playbackQueueController.clear();
            host.playbackUiState.queue.clear();
        } else {
            host.playbackQueueController.removeCommitted(removed.trackIds, removed.trackUris);
        }
        host.libraryState.tracks.removeIf(track -> removed.trackIds.contains(track.trackId));
        host.libraryState.favorites.removeAll(removed.trackUris);
        for (Playlist playlist : host.libraryState.playlists) {
            playlist.uris.removeAll(removed.trackUris);
        }
        host.libraryRepository.reindex();
        host.librarySnapshotApplier.rebuildDerivedAndRender();
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/LibraryMutationController.kt
package com.dumuzeyn.mp3player

import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException

/** Coordinates committed SQLite removals with the activity and Media3 projections. */
internal class LibraryMutationController(private val host: MainActivityCore) : AutoCloseable {
    private val executor = Executors.newSingleThreadExecutor()
    @Volatile private var closed = false

    fun removeTrack(track: Track?) {
        removeTrack(track, false)
    }

    fun removeDeletedFile(track: Track?) {
        removeTrack(track, true)
    }

    fun removeSource(source: LibrarySource?) {
        if (source == null) return
        execute { publish(PersistedFolderStore.forget(host, source)) }
    }

    fun clearLibrary() {
        execute { publish(PersistedFolderStore.clear(host)) }
    }

    override fun close() {
        closed = true
        executor.shutdownNow()
    }

    private fun removeTrack(track: Track?, fileWasDeleted: Boolean) {
        if (track == null) return
        execute {
            LibraryMutationStore(host).use { store ->
                publish(
                    if (fileWasDeleted) store.removeDeletedFile(track) else store.removeTrack(track),
                )
            }
        }
    }

    private fun execute(mutation: () -> Unit) {
        try {
            executor.execute {
                try {
                    mutation()
                } catch (error: RuntimeException) {
                    VoltuneLog.failure("library_mutation_failed", error)
                }
            }
        } catch (_: RejectedExecutionException) {
            // Activity is already closing.
        }
    }

    private fun publish(removed: RemovedLibraryItems) {
        removed.sources.forEach { source ->
            PersistedFolderStore.releaseReadPermission(host, source.asUri())
        }
        host.uiHandler.post { applyToUi(removed) }
    }

    private fun applyToUi(removed: RemovedLibraryItems) {
        if (closed) return
        if (removed.clearQueue) {
            host.playbackQueueController.clear()
            host.playbackUiState.queue.clear()
        } else {
            host.playbackQueueController.removeCommitted(removed.trackIds, removed.trackUris)
        }
        for (index in host.libraryState.tracks.lastIndex downTo 0) {
            if (host.libraryState.tracks[index].trackId in removed.trackIds) {
                host.libraryState.tracks.removeAt(index)
            }
        }
        host.libraryState.favorites.removeAll(removed.trackUris)
        host.libraryState.playlists.forEach { playlist ->
            playlist.uris.removeAll(removed.trackUris)
        }
        host.libraryRepository.reindex()
        host.librarySnapshotApplier.rebuildDerivedAndRender()
    }
}
